// tfx-cli/_build/lib/arguments.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var common = require("../lib/common");
var diskcache_1 = require("../lib/diskcache");
var path = require("path");
var qread = require("./qread");
var util_1 = require("util");
var fsUtils = require("./fsUtils");
var fs_1 = require("fs");
/**
 * Class that represents an argument with a value. Calling .val() will retrieve
 * the typed value, parsed from the givenValue (a string). If no givenValue
 * was provided, we will prompt the user.
 */
var Argument = /** @class */ (function () {
    function Argument(name, friendlyName, description, givenValue, hasDefaultValue, aliases, undocumented, promptDefault) {
        if (friendlyName === void 0) { friendlyName = name; }
        if (undocumented === void 0) { undocumented = false; }
        this.name = name;
        this.friendlyName = friendlyName;
        this.description = description;
        this.hasDefaultValue = hasDefaultValue;
        this.aliases = aliases;
        this.undocumented = undocumented;
        this.promptDefault = promptDefault;
        this.silent = false;
        if (typeof givenValue === "string") {
            this.givenValue = [givenValue];
        }
        else {
            this.givenValue = givenValue;
        }
        this.initialize();
    }
    /**
     * If this argument was given a default value:
     *   check the cache
     *     if it's there, set assignedValue to the getValue(cachedValue)
     *     else set assigned value to given default
     * If this argument was given a default value of null
     *   set null as the assignedValue
     * If this argument was not given any value
     *   check the cache
     *     if it's there, set assignedValue to cachedValue
     *
     * Promise is resolved after any values that need parsing are parsed,
     * and there are no more calls to the cache.
     */
    Argument.prototype.initialize = function () {
        var _this = this;
        var initPromise = Promise.resolve(null);
        if (this.assignedValue === undefined && (this.hasDefaultValue || this.givenValue === undefined)) {
            initPromise = getOptionsCache().then(function (cache) {
                var cacheKey = path.resolve().replace("/.[]/g", "-") +
                    "." +
                    common.EXEC_PATH.slice(0, common.EXEC_PATH.length - 1).join("/");
                var cachedValue = _.get(cache, cacheKey + "." + _this.name);
                var cachedValueStringArray;
                if (typeof cachedValue === "string") {
                    cachedValueStringArray = [cachedValue];
                }
                else if (_.isArray(cachedValue)) {
                    cachedValueStringArray = cachedValue;
                }
                if (cachedValue !== undefined) {
                    return _this.getValue(cachedValueStringArray).then(function (result) {
                        _this.initializeAssignedValue(result);
                    });
                }
                else if (_this.givenValue !== null && _this.givenValue !== undefined) {
                    return _this.getValue(_this.givenValue).then(function (result) {
                        _this.initializeAssignedValue(result);
                    });
                }
                else if (_this.givenValue === null) {
                    _this.initializeAssignedValue(null);
                }
            });
        }
        else if (this.assignedValue === undefined) {
            if (this.givenValue === null) {
                this.initializeAssignedValue(null);
            }
            else if (this.givenValue !== undefined) {
                initPromise = this.getValue(this.givenValue).then(function (result) {
                    _this.initializeAssignedValue(result);
                });
            }
        }
        this.initializePromise = initPromise;
        return initPromise;
    };
    Argument.prototype.initializeAssignedValue = function (val) {
        if (this.assignedValue === undefined) {
            this.assignedValue = val;
        }
    };
    /**
     * Override whatever exists and give this argument a value.
     */
    Argument.prototype.setValue = function (value) {
        this.assignedValue = value;
        this.initializePromise = Promise.resolve(null);
    };
    /**
     * Get the value of this argument by what was passed in. If nothing has
     * been passed in, prompt the user. The resulting promise is resolved
     * when a value is available.
     */
    Argument.prototype.val = function (noPrompt) {
        var _this = this;
        if (noPrompt === void 0) { noPrompt = false; }
        return this.initializePromise.then(function () {
            if (_this.assignedValue !== undefined) {
                return Promise.resolve(_this.assignedValue);
            }
            else {
                if (!noPrompt && !_this.undocumented) {
                    if (common.NO_PROMPT) {
                        throw new Error("Missing required value for argument '" + _this.name + "'.");
                    }
                    return qread.read(_this.name, _this.friendlyName, _this.silent, _this.promptDefault).then(function (answer) {
                        // Split answer into args, just as if they were passed through command line
                        var splitAnswer = answer.match(/".+?"|[^ ]+/g) || [""];
                        var answerArgs = splitAnswer.map(function (a) {
                            // trim quotes if needed
                            if (a.substr(0, 1) === '"' && a.substr(a.length - 1, 1) === '"') {
                                a = a.substr(1, a.length - 1);
                            }
                            return a;
                        });
                        return _this.getValue(answerArgs).then(function (result) {
                            _this.assignedValue = result;
                            _this.hasDefaultValue = false;
                            return result;
                        });
                    });
                }
                else {
                    return Promise.resolve(null);
                }
            }
        });
    };
    return Argument;
}());
exports.Argument = Argument;
/**
 * Argument that represents an array of comma-separated strings.
 */
var ArrayArgument = /** @class */ (function (_super) {
    __extends(ArrayArgument, _super);
    function ArrayArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ArrayArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, stripped;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        if (params.length === 1) {
                            stripped = params[0].replace(/(^\[)|(\]$)/g, "");
                            return [2 /*return*/, Promise.resolve(stripped.split(",").map(function (s) { return s.trim(); }))];
                        }
                        else {
                            return [2 /*return*/, Promise.resolve(params)];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return ArrayArgument;
}(Argument));
exports.ArrayArgument = ArrayArgument;
/**
 * Argument that represents a set of file paths.
 * @TODO: Better validation of valid/invalid file paths (FS call?)
 */
var FilePathsArgument = /** @class */ (function (_super) {
    __extends(FilePathsArgument, _super);
    function FilePathsArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FilePathsArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        return [2 /*return*/, Promise.resolve(params.map(function (p) { return path.resolve(p); }))];
                }
            });
        });
    };
    return FilePathsArgument;
}(Argument));
exports.FilePathsArgument = FilePathsArgument;
/**
 * Argument that represents a set of existing file paths
 */
var ExistingFilePathsArgument = /** @class */ (function (_super) {
    __extends(ExistingFilePathsArgument, _super);
    function ExistingFilePathsArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExistingFilePathsArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.getValue.call(this, argParams).then(function (paths) {
                        var existencePromises = [];
                        paths.forEach(function (p) {
                            var promise = fsUtils.exists(p).then(function (exists) {
                                if (!exists) {
                                    throw new Error("The file at path " + p + " does not exist.");
                                }
                                else {
                                    return p;
                                }
                            });
                            existencePromises.push(promise);
                        });
                        return Promise.all(existencePromises);
                    })];
            });
        });
    };
    return ExistingFilePathsArgument;
}(FilePathsArgument));
exports.ExistingFilePathsArgument = ExistingFilePathsArgument;
/**
 * Argument that represents a set of writable file paths.
 * Paths that refer to existing files are checked for writability
 * Paths that refer to non-existent files are assumed writable.
 */
var WritableFilePathsArgument = /** @class */ (function (_super) {
    __extends(WritableFilePathsArgument, _super);
    function WritableFilePathsArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WritableFilePathsArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.getValue.call(this, argParams).then(function (paths) {
                        var canWritePromises = [];
                        paths.forEach(function (p) {
                            var promise = fsUtils.canWriteTo(p).then(function (canWrite) {
                                if (canWrite) {
                                    return p;
                                }
                                else {
                                    throw new Error("The file at path " + p + " is not writable.");
                                }
                            });
                            canWritePromises.push(promise);
                        });
                        return Promise.all(canWritePromises);
                    })];
            });
        });
    };
    return WritableFilePathsArgument;
}(FilePathsArgument));
exports.WritableFilePathsArgument = WritableFilePathsArgument;
/**
 * Argument that represents a set of readable file paths
 */
var ReadableFilePathsArgument = /** @class */ (function (_super) {
    __extends(ReadableFilePathsArgument, _super);
    function ReadableFilePathsArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ReadableFilePathsArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.getValue.call(this, argParams).then(function (paths) {
                        var canReadPromises = [];
                        paths.forEach(function (p) {
                            var promise = fsUtils.fileAccess(p, fsUtils.R_OK).then(function (canRead) {
                                if (canRead) {
                                    return p;
                                }
                                else {
                                    throw new Error("The file at path " + p + " is not readable.");
                                }
                            });
                            canReadPromises.push(promise);
                        });
                        return Promise.all(canReadPromises);
                    })];
            });
        });
    };
    return ReadableFilePathsArgument;
}(ExistingFilePathsArgument));
exports.ReadableFilePathsArgument = ReadableFilePathsArgument;
/**
 * Argument that represents a set of existing directory file paths
 */
var ExistingDirectoriesArgument = /** @class */ (function (_super) {
    __extends(ExistingDirectoriesArgument, _super);
    function ExistingDirectoriesArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExistingDirectoriesArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, _super.prototype.getValue.call(this, argParams).then(function (paths) {
                        var isDirectoryPromises = [];
                        paths.forEach(function (p) {
                            var promise = util_1.promisify(fs_1.lstat)(p).then(function (stats) {
                                if (stats.isDirectory()) {
                                    return p;
                                }
                                else {
                                    throw new Error("The path " + p + " is not a directory.");
                                }
                            });
                            isDirectoryPromises.push(promise);
                        });
                        return Promise.all(isDirectoryPromises);
                    })];
            });
        });
    };
    return ExistingDirectoriesArgument;
}(ExistingFilePathsArgument));
exports.ExistingDirectoriesArgument = ExistingDirectoriesArgument;
/**
 * Argument that represents a boolean value.
 */
var BooleanArgument = /** @class */ (function (_super) {
    __extends(BooleanArgument, _super);
    function BooleanArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * If a value is given, parse it and cache the value.
     */
    BooleanArgument.prototype.initialize = function () {
        var _this = this;
        this.initializePromise = Promise.resolve(null);
        if (this.givenValue !== undefined) {
            if (this.givenValue === null) {
                this.assignedValue = false;
                this.initializePromise = Promise.resolve(null);
            }
            else {
                this.initializePromise = this.getValue(this.givenValue).then(function (result) {
                    _this.assignedValue = result;
                });
            }
        }
        return this.initializePromise;
    };
    /**
     * If there is no argument to this option, assume true.
     */
    BooleanArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, yes, no;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        if (params.length === 1) {
                            yes = ["true", "1", "yes", "y"].indexOf(params[0].toLowerCase()) >= 0;
                            if (yes) {
                                return [2 /*return*/, Promise.resolve(true)];
                            }
                            no = ["false", "0", "no", "n"].indexOf(params[0].toLowerCase()) >= 0;
                            if (no) {
                                return [2 /*return*/, Promise.resolve(false)];
                            }
                            throw new Error("'" + params[0] + "' is not a recognized Boolean value.");
                        }
                        else if (params.length === 0) {
                            return [2 /*return*/, Promise.resolve(true)];
                        }
                        else {
                            throw new Error("Multiple values provided for Boolean Argument " + this.name + ".");
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return BooleanArgument;
}(Argument));
exports.BooleanArgument = BooleanArgument;
/**
 * Argument that reprents an int value.
 */
var IntArgument = /** @class */ (function (_super) {
    __extends(IntArgument, _super);
    function IntArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IntArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, parseResult;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        if (params.length === 1) {
                            parseResult = parseInt(params[0], 10);
                            if (isNaN(parseResult)) {
                                throw new Error("Could not parse int argument " + this.name + ".");
                            }
                            return [2 /*return*/, Promise.resolve(parseResult)];
                        }
                        else if (params.length === 0) {
                            throw new Error("No number provided for Int Argument " + this.name + ".");
                        }
                        else {
                            throw new Error("Multiple values provided for Int Argument " + this.name + ".");
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return IntArgument;
}(Argument));
exports.IntArgument = IntArgument;
/**
 * Argument that reprents a float value.
 */
var FloatArgument = /** @class */ (function (_super) {
    __extends(FloatArgument, _super);
    function FloatArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FloatArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, parseResult;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        if (params.length === 1) {
                            parseResult = parseFloat(params[0]);
                            if (isNaN(parseResult)) {
                                throw new Error("Could not parse float argument " + this.name + ".");
                            }
                            return [2 /*return*/, Promise.resolve(parseResult)];
                        }
                        else if (params.length === 0) {
                            throw new Error("No number provided for Float Argument " + this.name + ".");
                        }
                        else {
                            throw new Error("Multiple values provided for Float Argument " + this.name + ".");
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return FloatArgument;
}(Argument));
exports.FloatArgument = FloatArgument;
/**
 * Argument that represents a block of JSON.
 * Note: This class must be extended with a concrete type before its constructor
 * function can be referenced. See exec/extensions/default.ts for an example.
 */
var JsonArgument = /** @class */ (function (_super) {
    __extends(JsonArgument, _super);
    function JsonArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JsonArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, info;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        try {
                            return [2 /*return*/, Promise.resolve(JSON.parse(params.join(" ")))];
                        }
                        catch (parseError) {
                            info = parseError.stack || parseError.message;
                            throw new Error("Failed to parse JSON argument " + this.name + ". Info: " + info);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return JsonArgument;
}(Argument));
exports.JsonArgument = JsonArgument;
/**
 * Argument that represents a string. Multiple values are joined together
 * by a single space.
 */
var StringArgument = /** @class */ (function (_super) {
    __extends(StringArgument, _super);
    function StringArgument() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StringArgument.prototype.getValue = function (argParams) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!Array.isArray(argParams)) return [3 /*break*/, 1];
                        _a = argParams;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, argParams];
                    case 2:
                        _a = _b.sent();
                        _b.label = 3;
                    case 3:
                        params = _a;
                        return [2 /*return*/, Promise.resolve(params.join(" "))];
                }
            });
        });
    };
    return StringArgument;
}(Argument));
exports.StringArgument = StringArgument;
/**
 * Argument that represents a string, however, if we ever have to
 * prompt the user for the value of this argument, we do not echo
 * out the value as it is typed. Good for passwords, tokens, etc.
 */
var SilentStringArgument = /** @class */ (function (_super) {
    __extends(SilentStringArgument, _super);
    function SilentStringArgument() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.silent = true;
        return _this;
    }
    return SilentStringArgument;
}(StringArgument));
exports.SilentStringArgument = SilentStringArgument;
function getOptionsCache() {
    var cache = new diskcache_1.DiskCache("tfx");
    return cache.itemExists("cache", "command-options").then(function (cacheExists) {
        var existingCache = Promise.resolve("{}");
        if (cacheExists) {
            existingCache = cache.getItem("cache", "command-options");
        }
        return existingCache.then(function (cacheStr) {
            try {
                return JSON.parse(cacheStr);
            }
            catch (ex) {
                return {};
            }
        });
    });
}
exports.getOptionsCache = getOptionsCache;
