// tfx-cli/_build/lib/trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var colors = require("colors");
var os = require("os");
var debugTracingEnvVar = process.env["TFX_TRACE"];
exports.traceLevel = debugTracingEnvVar ? 2 /* Debug */ : 1 /* Info */;
exports.debugLogStream = console.log;
function println() {
    info("");
}
exports.println = println;
function error(msg) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    log("error: ", msg, colors.bgRed, replacements, console.error);
}
exports.error = error;
function success(msg) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    log("", msg, colors.green, replacements);
}
exports.success = success;
function info(msg) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    if (exports.traceLevel >= 1 /* Info */) {
        log("", msg, colors.white, replacements);
    }
}
exports.info = info;
function warn(msg) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    log("warning: ", msg, colors.bgYellow.black, replacements);
}
exports.warn = warn;
function debugArea(msg, area) {
    debugTracingEnvVar = process.env["TFX_TRACE_" + area.toUpperCase()];
    if (debugTracingEnvVar) {
        log(colors.cyan(new Date().toISOString() + " : "), msg, colors.grey, [], exports.debugLogStream);
    }
    debugTracingEnvVar = process.env["TFX_TRACE"];
}
exports.debugArea = debugArea;
function debug(msg) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    if (exports.traceLevel >= 2 /* Debug */) {
        log(colors.cyan(new Date().toISOString() + " : "), msg, colors.grey, replacements, exports.debugLogStream);
    }
}
exports.debug = debug;
function log(prefix, msg, color, replacements, method) {
    if (method === void 0) { method = console.log; }
    var t = typeof msg;
    if (t === "string") {
        write(prefix, msg, color, replacements, method);
    }
    else if (msg instanceof Array) {
        msg.forEach(function (line) {
            if (typeof line === "string") {
                write(prefix, line, color, replacements, method);
            }
        });
    }
    else if (t === "object") {
        write(prefix, JSON.stringify(msg, null, 2), color, replacements, method);
    }
}
function write(prefix, msg, color, replacements, method) {
    if (method === void 0) { method = console.log; }
    var toLog = format.apply(void 0, [msg].concat(replacements));
    toLog = toLog
        .split(/\n|\r\n/)
        .map(function (line) { return prefix + line; })
        .join(os.EOL);
    method(color(toLog));
}
function format(str) {
    var replacements = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        replacements[_i - 1] = arguments[_i];
    }
    var lcRepl = str.replace(/%S/g, "%s");
    var split = lcRepl.split("%s");
    if (split.length - 1 !== replacements.length) {
        throw new Error("The number of replacements (" +
            replacements.length +
            ") does not match the number of placeholders (" +
            (split.length - 1) +
            ")");
    }
    var resultArr = [];
    split.forEach(function (piece, index) {
        resultArr.push(piece);
        if (index < split.length - 1) {
            resultArr.push(replacements[index]);
        }
    });
    return resultArr.join("");
}
exports.format = format;
