// tfx-cli/_build/exec/extension/isvalid.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var args = require("../../lib/arguments");
var colors = require("colors");
var extBase = require("./default");
var publishUtils = require("./_lib/publish");
var trace = require("../../lib/trace");
function getCommand(args) {
    return new ExtensionIsValid(args);
}
exports.getCommand = getCommand;
var ExtensionIsValid = /** @class */ (function (_super) {
    __extends(ExtensionIsValid, _super);
    function ExtensionIsValid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.description = "Show the validation status of a given extension.";
        _this.serverCommand = true;
        return _this;
    }
    ExtensionIsValid.prototype.setCommandArgs = function () {
        _super.prototype.setCommandArgs.call(this);
        this.registerCommandArgument("version", "Extension version", "Specify the version of the extension of which to get the validation status. Defaults to the latest version.", args.StringArgument, null);
        this.registerCommandArgument("serviceUrl", "Market URL", "URL to the VSS Marketplace.", args.StringArgument, extBase.ExtensionBase.getMarketplaceUrl);
    };
    ExtensionIsValid.prototype.getHelpArgs = function () {
        return ["publisher", "extensionId", "vsix", "version"];
    };
    ExtensionIsValid.prototype.exec = function () {
        return __awaiter(this, void 0, void 0, function () {
            var galleryApi, extInfo, version, sharingMgr, validationStatus;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getGalleryApi()];
                    case 1:
                        galleryApi = _a.sent();
                        return [4 /*yield*/, this.identifyExtension()];
                    case 2:
                        extInfo = _a.sent();
                        return [4 /*yield*/, this.commandArgs.version.val()];
                    case 3:
                        version = _a.sent();
                        sharingMgr = new publishUtils.SharingManager({}, galleryApi, extInfo);
                        return [4 /*yield*/, sharingMgr.getValidationStatus(version)];
                    case 4:
                        validationStatus = _a.sent();
                        return [2 /*return*/, validationStatus];
                }
            });
        });
    };
    ExtensionIsValid.prototype.friendlyOutput = function (data) {
        if (data === publishUtils.GalleryBase.validated) {
            trace.info(colors.green("Valid"));
        }
        else if (data === publishUtils.GalleryBase.validationPending) {
            trace.info(colors.yellow("Validation pending..."));
        }
        else {
            trace.info(colors.red("Validation error: " + data));
        }
    };
    ExtensionIsValid.prototype.jsonOutput = function (data) {
        var result = {
            status: "error",
        };
        if (data === publishUtils.GalleryBase.validationPending) {
            result.status = "pending";
        }
        else if (data === publishUtils.GalleryBase.validated) {
            result.status = "success";
        }
        else {
            result.message = data;
        }
        console.log(JSON.stringify(result, null, 4));
    };
    return ExtensionIsValid;
}(extBase.ExtensionBase));
exports.ExtensionIsValid = ExtensionIsValid;
